Adding option to zero out heterogeneity in upper X m, where X can be …

jaustermann opened the pull request:
…specified in input file

I think this would be a useful small addition that I would like to use in the S40RTS cookbook that I'm working on.

tjhei commented:
Can you please fix the indentation?

bangerth commented on source/initial_conditions/S40RTS_perturbation.cc:
```diff
@@ -625,6 +631,10 @@ namespace aspect
+          prm.declare_entry ("Remove temperature heterogeneity down to specified depth", "-10",
+                             Patterns::Double (),
+                             "This will set the heterogeneity prescribed by S20RTS or S40RTS to zero "
+                             "down to the specified depth.");
```
Can you state the units of this parameter (I assume meters)? Also, what does the default value of -10 mean: a *height* of ten meters? If so, why?

If you want the default to be to not zero out any heterogeneities, then setting it to some very large negative value would be the right thing. Zero would also be a valid value.

jaustermann replied:
I would like to set it to zero, but didn't because I thought this might cause a problem if the depth of a given point isn't exactly 0 but slightly smaller (due to numerical error). Can that not happen? The value of -10 is random, could also be a very large negative value.

bangerth replied:
I suspect that that can indeed happen. I think you probably want to set it to a very large negative value, e.g., `-std::numeric_limits<double>::max()` and its corresponding string representation which you can get via `boost::lexical_cast<std::string>(std::numeric_limits<double>::max())`.

bangerth commented:
Please fix the indentation as shown in http://www.math.clemson.edu/~heister/aspect-astyler-logs/ec2e7b866b3bb3cb44cdd5c653a4c293570a0cbd/log .

Can you add an entry to doc/modules/changes.h? (You may need to rebase to master to make it possible to merge in that case.)

jaustermann commented:
I messed this up, sorry! I was trying to send the pull request from a different branch. Let me see how I can fix this.

jaustermann commented:
@bangerth I updated this pull request: Fixed the astyle and default value. I also expanded the documentation a little bit. I messed up the git push though, which led to a series of commits, so sorry! I can talk to you or @tjhei to see how to deal with this.

bangerth commented on include/aspect/initial_conditions/S40RTS_perturbation.h:
```diff
@@ -104,9 +104,12 @@ namespace aspect
          */
         double vs_to_density;
         double thermal_alpha;
+        double no_pertubation_depth;
```
typo

bangerth commented on source/initial_conditions/S40RTS_perturbation.cc:
```diff
@@ -315,10 +315,16 @@ namespace aspect
+      // zero out heterogeneity down to a specified depth
+      const double depth = this->get_geometry_model().depth(position);
+      if (depth <= no_pertubation_depth)
+        temperature_perturbation = 0.0;
```
do
```
  double temp_perturb;
  if (...)
    temp_perturb = ...;
  else
    temp_perturb = ...;
```

bangerth commented on source/initial_conditions/S40RTS_perturbation.cc:
```diff
@@ -625,6 +631,14 @@ namespace aspect
                              Patterns::Double (0),
                              "The reference temperature that is perturbed by the spherical "
                              "harmonic functions. Only used in incompressible models.");
+          prm.declare_entry ("Remove temperature heterogeneity down to specified depth", boost::lexical_cast<std::string>(std::numeric_limits<double>::max()),
```
minus max
